proposals/0000-updated-partial-type-sigs.rst

```diff
@@ -0,0 +1,497 @@
+
+Unresolved Questions
+--------------------
+* Is this really the best syntax? I am uncomfortable at stealing both underscored-identifiers and ``::?``.
+  How painful is it to do so?
+
+* Is this design too elaborate? I have a tendency to build elaborate but expressive edifices. Perhaps
+  there is a sweet spot closer to the ground here.
+  
+Further examples
+----------------
+
+.. _below:
+  
+
+* Let's dive deeper into this example::
+
+    silly ::? Proxy a -> ()
+    silly (_ :: Proxy @Bool _) = ()
+
+  Actually, let's first consider something very closely related::
+
+    sillier :: Proxy a -> ()
+    sillier (_ :: Proxy @Bool _) = ()
+
+  This definition of ``sillier`` is rejected. That's because GHC processes its type signature in isolation.
+  GHC sees that we wish to quantify over ``a``. After kind-checking, the kind of ``a`` is utterly unconstrained.
+  Thus, GHC infers ``sillier :: forall {k} (a :: k). Proxy @k a -> ()``. All of this has happened *without*
+  looking at the definition of ``sillier``. When GHC does look at that definition, it is rejected, as
+  the argument to ``sillier`` has type ``Proxy @k a``, not ``Proxy @Bool a``. This is the kind-level
+  equivalent of something like ::
+
+    silliest :: a -> a
+    silliest True = False
+
+  In both ``sillier`` and ``silliest``, the type signature is *more general* than the definition. It's
+  just that ``sillier`` is harder to see that.
+
+  Returning to ``silly``, with the partial type signature marker ``::?``, GHC will *not* kind-generalize
+  the type. It will effectively infer ``silly ::? forall (a :: _k). Proxy @_k a -> ()``, where ``_k``
+  behaves like a named wildcard. (I say "behaves like" because named wildcards become unification variables
+  internally; in this case, the kind of ``a`` really would just be a unification variable.) Now, when
+  checking the definition of ``silly``, GHC is free to discover that ``_k`` should be ``Bool``, and all is
+  well.
+
+  Note that I did *not* mean to write ``silly ::? Proxy _a -> ()``. I want ``a`` to be a skolem here. It's
+  ``a``\s *kind* that I want not to be a skolem.
+
+  Thus, ``silly`` is silly only because it is contrived, not because it is wrong.
+
+* Here are some examples around the use of elisions:
+
+  1. ::
+
+       f :: _ -> _
+
+     This would mean the same as ``f :: a -> b``.
+
+  2. ::
+
+       f :: forall a. a -> _
+
+     This would mean the same as ``f :: forall a. forall b. a -> b`` (the second ``forall`` prevents
+     ``b`` from being in scope in the defined term). I'm sure you now ask: "Why the unrequested
+     generalization?" Because this is a complete type signature, not a partial one. The new
+     generalization behavior affects only *partial* type signatures. So, ::
+
+       f ::? forall a. a -> _
+
+     would use the definition for ``f`` to determine the result type, but that type could not
+     mention a type variable other than ``a`` (assuming we're at top-level).
```

Comment:
Why do partial signatures affect the generalization strategy? That doesn't seem to be explained in the proposal (nor does the need for a new `x ::? t` form for that matter), but it seems quite counterintuitive. Wouldn't it make more sense to say that the generalization strategy depends on the thing being generalized? Then we have three cases:

1. plain type variables: implicitly generalized unless there's an explicit forall.
2. type wildcards: generalized if there's a forall elision (though see my comment above).
3. type elision: implicitly generalized always.

Reply:
I'm OK with this change. It does seem simpler.